### ISSUE TITLE
Smoothen reward distribution to linear function

### ISSUE THREAD

**Issue by litrbooh** [GolosChain/golos#283]
В рамках текущего ХФ необходимо сгладить функцию распределения выплат до линейной (с возможными нюансами).

Напоминаю, что изменение квадратичной функции было обещано ещё до старта Голоса - более года назад:
![default](https://user-images.githubusercontent.com/22373216/32914513-da398f2c-cb26-11e7-9eba-a72db0d42e50.png)

Основная задача сглаживания - сделать распределение выплат справедливым, чтобы каждый мог распределять пул **прямо пропорционально своему весу**. Предложенная ранее логарифмическая функция чуть более справедлива, чем квадратичная, но абсолютно не обоснованно усиление веса и без того крупных держателей. Как средство выделения **"качественного"** контента это не работает - уже проверено. 

Напоминаю, что  стим прежде чем перейти на линейную выплату провел эксперимент: https://steemit.com/steem/@abit/some-data-about-the-whales-no-voting-experiment-and-hard-fork-18

Вот здесь оникс представил расчеты с тем, как изменятся выплаты на примере одного дня:
https://golos.blog/ru--delegat/@on1x/reward-system

Сама таблица: https://docs.google.com/spreadsheets/d/1mBEpG2a_LYzLgGy8J5rxa8d1z4hJdkaeU2g8S_Xrtg8/edit#gid=1853920859

Основная идея  в том, что сотни новых пользователей приходящих на голос получали хоть какие-то копейки и видели возможности роста и накоплении веса. Смотрите разницу внизу:
![default](https://user-images.githubusercontent.com/22373216/32914967-66406f30-cb28-11e7-9115-cfc6489239c5.png)

Наглядно показано, что тысячи пользователей вместо нулей начнут получать единицы рублей и распределять копейки- это достаточный стимул, чтобы стимулировать их оставаться на платформе.

Когда будет у нас мощный средний класс и не будет крупных китов, тогда подумаем о вашем логарифме.

**Comment by asuleymanov** [comment]
Я ничего сказать не хочу но линейная функция и логарифмическая это разные вещи.

**Comment by 21xhipster** [comment]
Golos Core линейную функцию имплементировать не будет

**Comment by On1x** [comment]
Ок, не линейную, а другую, но сгладить квадратичную просто необходимо.

**Comment by On1x** [comment]
Я бы предложил рассмотреть компромисс до следующего хардфорка rshares + (rshares*rshares)/3

**Comment by 21xhipster** [comment]
O(n log n) ?

**Comment by 21xhipster** [comment]
![jIGhf.png](https://shipusercontent.com/babbe17aa4420be1430e5a369b99ff00/jIGhf.png)

**Comment by On1x** [comment]
Да, почему бы и нет. Это внедрить минут на 10 - несколько строчек кода, так как финальный подсчет суммарного rshares в 1 функции.

**Comment by On1x** [comment]
Я только за параметеризацию сети в зависимости от голосования, но сейчас часть делегатов встанут в позу - "вместе с делегированием квадрат принимать не будем". Не раз слышал эту позицию. Почему не пойти на встречу? Это не так сложно.

**Comment by 21xhipster** [comment]
Стоит прочитать [эту серию статей](https://the-steem-reward-system-part-1-v-shares/) чтобы понять что такие штуки за 10 минут на коленке не делаются. Толь, это тебе не фронтенд лабать. Если честно я против добавления нового скоупа в этот хадрфорк кроме фиксов того, что надо пофиксить.

**Comment by 21xhipster** [comment]
> "вместе с делегированием квадрат принимать не будем". Не раз слышал эту позицию. Почему не пойти на встречу? Это не так сложно.

Я если честно не понял это вопрос (и если да то к кому) или утверждение?

**Comment by tomarcafe** [comment]
а картинка пёстрая она к чему? как она относится вообще к теме?

**Comment by 21xhipster** [comment]
Это меню, Леш

**Comment by tomarcafe** [comment]
ссылку на серию статей поправь, ведёт в пустоту.

**Comment by tomarcafe** [comment]
ну так там написано:  O(n^2) - Horrible. O(n) - Fair.
к чему меню-то, Дим?

**Comment by tomarcafe** [comment]
что такое Operations, что тауое Elements? или ты просто показать скилл вставляния картинок демонстрируешь?

**Comment by 21xhipster** [comment]
А я легенду не подметил :-) Она не совсем в нашу историю укладывается :_)

**Comment by On1x** [comment]
Дима, я согласен, что за 10 минут на коленке не делаются, но человек скриншот прислал, что это планировалось ранее. Костя мне сказал, что у вас делается аналитика. Достаточно выложить ее в паблик и обсудить.

По поводу позиции части делегатов - несмотря на то, что я уделяю все время "лабанью" фронта, нахожусь в курсе настроений в сообществе https://goldvoice.club/@arcange/byt-khardforku-ili-ne-byt-vot-v-chem-vopros/
Тот же Аркандж довольно четко высказал свою позицию - цитирую:
>Я не стану принимать новый хардфорк с квадратичной моделью.

И не он один. Если у вас есть аналитика, которую делают не 1 месяц, то в чем проблема обсудить варианты, проголосовать и внедрить ее за 10 минут (в коде это реально одна функция https://github.com/GolosChain/golos/blob/v0.16.4/libraries/chain/database.cpp#L2423 )?

**Comment by 21xhipster** [comment]
@lomashuk Давай спросим у него какая аналитика имелась ввиду

**Comment by tomarcafe** [comment]
Наверное, речь шла про ту аналитику, которую Марина на митинге упоминала?

**Comment by On1x** [comment]
Дима, есть, конечно, вероятность, что я что-то недопонял. Возможно меня ввели в заблуждении прошлые статьи (их было несколько, в самом issue есть скриншот годичной давности), но создается впечатление, что какой-то анализ уже есть. Если посты только подготавливаются, то лучше поспешить, иначе получим ситуацию когда топ делегаты будут накладывать вето на ХФ.

Это стандартная ситуация, для это и создан механизм консенсуса. Буду рад, если ты напишешь, была ли какая-то аналитика и моделирование в прошлом, или нет.

**Comment by 21xhipster** [comment]
@On1x Вот неплохие пруфы на тему квадратичного голосования:
- https://arxiv.org/pdf/1409.0264.pdf
- https://www.sss.ias.edu/files/pdfs/Rodrik/workshop%2014-15/Weyl-Quadratic_Voting.pdf

Если совсем по простому (в тему квадратичного голосования) то давай объясню на пальцах:
Представь что есть инфляция за которой все охотятся. С линейной функцией каждый может получать свою часть инфляции в независимости от производимого контента. В результате не будет происходить редистрибьюции или её эффект будет очень слабым. Такой механизм будет только усугублять эгоистичное поведение, а не работу на общее благо. Квадратичная функция в нашем случае решает [традиционную трагедию общин](https://ru.wikipedia.org/wiki/%D0%A2%D1%80%D0%B0%D0%B3%D0%B5%D0%B4%D0%B8%D1%8F_%D0%BE%D0%B1%D1%89%D0%B8%D0%BD) на корню, т.к. получение выплат в такой системе возможно только в случае если произведено не персональное, но благо общественное. Без квадратичной функции экономическая система достаточно быстро сколапсирует. Те кто этого не понимают имеют возможность поэкспериментировать за свой счет на сети, которую создадут сами. Надеюсь это ясно.

Что касается вот этого:

> Я не стану принимать новый хардфорк с квадратичной моделью

Моя трактовка сводится к тому что делегат пренебрегает условиями первоначального договора и препятствует развитию сети. Если уважаемый аркандж будет такой один, то это не проблема. Если таких делегатов будет больше 5 то сеть застрянет в свое развитии до тех пор пока органическим образом делегаты не договорятся принять какой-то код. Я считаю что подобное упражнение многому научит сеть, и мы обязательно должны пройти через подобный кризис, чтобы двигаться дальше.

**Comment by On1x** [comment]
Спасибо за развернутый ответ. Пойду изучать доки.

По поводу делегатов. @VIM-Arcange может сам выразить свою позицию. Думаю, вполне возможна ситуация, что есть конкретное "против" квадратичности, чем "за" линейность. Вот эту ситуацию и надо разрешить.

**Comment by 21xhipster** [comment]
Толь, еще раз. За линейность в публичном сетапе не может быть ничего. Это тупо не жизнеспособно с точки зрения теории игр. Думаю пройдет максимум 36 месяцев прежде чем Стим сварится от жадности и алчности сообщества которое воспитает за это время его текущая экономика. Глупо опираться на неопытность ребят, которым достался от Дена этот проект. Компромис может быть где-то посередине между n2 и n. И то желательно до тех пора пока сообщество не поумнеет.

**Comment by On1x** [comment]
Да, я понял, что линейность нельзя внедрять по ряду причин. Про это и нет разговора (я все понял, серьезно). Но людям надо показать и объяснить "как мы хотим" и "почему". Мне нравится идея O(n log n). Почему бы не рассмотреть ее внедрение прямо сейчас? Тем более если ее изучали до этого и присматривались к ней?

**Comment by 21xhipster** [comment]
Можно, если @Nemo1369 и @kotbegemot скажут что они уверены, что можно безопасно её подмахнуть. Но в итоге эта хотелка может вылиться в еще целый месяц. Может быть мы примем текущий хардфорк с тем функционалом который доставлен, а потом начнем более короткие итерации? Я не против, если первой итерацией будет как раз эта штука.

**Comment by asuleymanov** [comment]
А может посмотреть с другой стороны. Может стоит сделать что нибудь не с самой [функцией подсчета](https://github.com/GolosChain/golos/blob/v0.16.4/libraries/chain/database.cpp#L2423).
Может стоит что нить сделать с выставление rshares к примеру rshares^1/2.

**Comment by tomarcafe** [comment]
Думаю, текущий хардфорк будет принят только после детального разбора по косточкам.
Всплывший функционал сброса ключей заставляет задуматься о многом. Говорить, что это из добрых намерений, можно, но поведение КФ эту гипотезу не подтверждает.
Неожиданное пропадание прошлогодних видео - тоже не подтверждает.

**Comment by On1x** [comment]
@21xhipster  Я пока не вижу "опасности" ее включения. Там тупо в коде обход постов и прогон через функцию суммарного rshares. Было бы отлично услышать их мнение.

@asuleymanov точно нет. Это изменения, которые могут повилять на всю экономику проекта. Поэтому обсуждать это "между делом" не стоит.

**Comment by sanchomalyutin** [comment]
@21xhipster

**Comment by On1x** [comment]
@sanchomalyutin можно не оффтопить и не передергивать. Речь про внутреннюю экономику серьезного проекта. Там цена ошибки гораздо выше, чем на каком-то ни было фронтенде.

**Comment by VIM-Arcange** [comment]
@On1x @21xhipster 

> I will not accept a new hardcore with a quadratic model

Indeed, I wrote this. I have to admit my  sentence is not properly formulated. I should have written:
> I will not approve the new hardcore with **both quadratic and power delegation**

This is because with this combination, some could game the reward pool by concentrating voting power, and Marina acknowledged this is an known issue.

The problem is that I really want power delegation, in order to be able to support projects.
Let's find something else than the binary situation "quadratic=no / Linear=yes"

**Comment by litrbooh** [comment]
@21xhipster 
Можно узнать почему ты закрыл ишью? Ты зам. директора Голоса, капитан блок-чейна или командир делегатов? Или ты закрыл, потому что ты из КФ, а только КФ в Голосе что-то решает?

Если последнее, то к чему тогда опус про DPOS и сообщество?
![default](https://user-images.githubusercontent.com/22373216/32941731-cc9eb7cc-cb97-11e7-81ac-4a1005238ce9.png)

Сообщество и несколько топ делегатов уже высказалось за линейную функцию, и тут ты приходишь и закрываешь ишью. А каком тогда ДПОС и сообществе идет речь?

Что касается сказок про квадратичность и сказок, про то, как она помогает выбрать качественный контент и самоорганизоваться, то мы уже это на голосе 3 недели наблюдаем, как в топе висят посты с откатами китам за апвоуты. 

Квадратичностью вы только стимулируете объединяться в группы для взаимного отката. При линейности в этом просто не будет смысла.

**Comment by Nemo1369** [comment]
Why it is closed?

**Comment by bitfag** [comment]
As I said before, there are no sense to use golos for small groups, communities or apps. Smaller groups always will get much lower rewards than dominating group. Remember why you forked golos at first place? Quadratic rewards, no?

**Comment by litrbooh** [comment]
@Nemo1369 @kotbegemot насколько трудоемко изменить функцию распределения в рамках текущего ХФ?

**Comment by t3ran13** [comment]
@On1x @21xhipster 

вариант толи "rshares + (rshares*rshares)/3"
можно принимать без опаски. он лишь увеличит незначительно влияние мальков, но при этом квадрат остается квадратом (формула квадратичной функции a*x^2 + b*x +c)

вот пример секущего квадрата (буду показывать для СГ, числа меньше чем шары, удобнее, для шар просто будет прямопорпорционально или нулей больше если попростому)
1 СГ ~ 1^2=1
10 СГ ~ 10^2=10
100 СГ ~ 100^2=10000
1 000 000 СГ  ~ 1 000 000^2=10^12


для формулы толика
1 СГ ~ 1+1^2/3 = 1,33 (было 1)
10 СГ ~ 10+10^2/3 = 43,33 (было 100)
100 СГ ~ 100 + 100^2/3 = 3433,33 (было 10 000)
1 000 000  СГ ~ 1 000 000 + 1 000 000^2/3 = 3,33^11 (10^12)

вот блин, выходит что хреновый вариант. сила апвота кита фактически не меняется, а вот средний класс начинает иметь меньше силы... мдя, реально, не следует бездумно лезть даже в код 1 функции

**Comment by Chiliec** [comment]
> (было 100)

10 же было?

**Comment by t3ran13** [comment]
но можно перестраховаться и сделать хотя бы "rshares + (rshares\*rshares)/2"

**Comment by On1x** [comment]
@t3ran13 тут идея как раз, что в формуле x+x*x/n — переменную n можно менять как угодно, подводя ближе в сторону к квадратичности или к линейности.

**Comment by Chiliec** [comment]
@On1x а кто это должен решать? Да и кто бы не решил, будет очередная волна постов и недовольства...

**Comment by t3ran13** [comment]
@On1x 
ну ты загнул, но не суть,ты прав что коэффициентами можно управлять. и вот для старта сделать хотя бы самый простой случай)

@Chiliec  вопрос не в том, что бы сейчас кто-то взял на себя бремя ответсвенности, вопрос в том, чб сейчас прийти к такой функции, которая бы устраивала всех)

**Comment by asuleymanov** [comment]
Думаю да сейчас можно применить какой то стандартный коэффициент. Типа 2 или 3 ну или 5 на худой конец.

**Comment by Chiliec** [comment]
Никогда не будет такой функции, которая бы устраивала всех. Надо сразу делегатам отдать, пускай настраивают :)

**Comment by litrbooh** [comment]
@21xhipster  @Nemo1369 @kotbegemot 

Немного с графикой
https://golos.io/ru--golos/@litrbooh/lineinost-vs-kvadratichnost-glazami-gumanitariya

**Comment by Nemo1369** [comment]
@21xhipster @On1x @litrbooh Seems like I would be able to publish a feature-branch with witness-votable author and curation reward curves. The only stuff to discuss is about the linear distribution function. I think @21xhipster would be definitely against it.

**Comment by litrbooh** [comment]
@Nemo1369
Давайте обсудим кривую с делегатами. Я свои аргументы привел, пока ничего другого я не видел. Если кто-то готов предложить свой вариант и объяснить чем он лучше линейности, тот пусть сделает это.


 Кроме того, надо учитывать эксперименты, которые проводил Steem.

**Comment by gropox** [comment]
Надо проголосовать тут, кто за линейность, за квадратичность или что то еще. Большинством принять.

**Comment by gropox** [comment]
Линейность

**Comment by gropox** [comment]
Квадратичность

**Comment by gropox** [comment]
Ничего не менять и ждать результатов анализа

**Comment by t3ran13** [comment]
Я против крайностей - квадрата и линейности. Мне нравится их одновременная комбинация в варианте @on1x

**Comment by litrbooh** [comment]
@gropox , тут не все делегаты, но можно попробовать

**Comment by Nemo1369** [comment]
@t3ran13 @gropox @litrbooh @21xhipster Guys, I did not meant the discussion of a final reward curve selection. I've meant some discussion about the the logarithmic curve representation.

**Comment by litrbooh** [comment]
@Nemo1369 
Я не очень понимаю, почему вдруг мы должны обсуждать логарифмическую кривую? Никто НИ разу в блок-чейн Голоса не предлагал ввести логарифмическую кривую. Откуда вообще это всплыло?

**Comment by Nemo1369** [comment]
@litrbooh Yeah, seems like a duplicate. Closed that one. Let is decide the final reward curves representations set.

**Comment by litrbooh** [comment]
@Nemo1369 
Как вы видите механизм окончательного решения по форме кривой?

**Comment by t3ran13** [comment]
Ну это явно не так круто чем квадрат, но лучше прямой. Но все равно литрбух задал хороший вопрос

**Comment by Chiliec** [comment]
Ребят, а почему нельзя взять в формулу дополнительные параметры, вроде кармы, нормализовать её и использовать это значение в любой формуле? 
Карма вроде не так просто накручивается, надо не просто накупить токенов, но и правда поучаствовать в жизни платформы. 
Активные авторы получили бы преимущество при распределении пула, что вроде неплохо.

**Comment by litrbooh** [comment]
@Chiliec потому что у нас ТОП-5 аккаунтов с репутацией, это или технические аккаунты еее, или пользователи публикующие в основном статистику. В общем, весьма сомнительный критерий полезности на платформе.

**Comment by Chiliec** [comment]
Будет не важно кто там в топе после нормализации, те кто рядом с топом получат приблизительно такие же значения. Кроме того это лишь дополнительный критерий, дающий небольшой, но существенный бонус для распределения.

Как человеку с маленькой кармой мне это не нравится, но я считаю что так было бы лучше для платформы, если квадрата не станет и ботоводить станет выгодно.

**Comment by t3ran13** [comment]
Идея не плохая, вообще лучше с переобуванием не спешить, а то можно натворитьь делов

**Comment by 21xhipster** [comment]
@Chiliec Это дело вводить в формулу новые переменные. Возраст аккаунта, например, если мы хотим простимулировать новичков. Карму, если мы хотим простимулировать более авторитетных авторов. Но @t3ran13 прав. Все работы по изменению кривой (в том числе затраты на рисерч) уйдут в разработку после уже разработанных и внедренных изменений.

**Comment by litrbooh** [comment]
опять кривая ушла, даже не на 0.2.1, а на 0.3.0 - это не серьезно

**Comment by t3ran13** [comment]
если честно, голос превратился в такое унылое говно из-за откатов, что ппц. 

Кривую выплат исправить на какую-то не квадратичную, уже и похер какую, можно прямо сегодня не дожидаясь основного хф, просто накотив изменения на текущий. там 1 формулу пофиксить

Это все можно закодить за час и накотить на следующие сутки, так что давайте мельтешить

Мы сейчас не авторов развиваем, а какой-то институт барыг на голосе создаем.

@Nemo1369  скажи в чем проблемы в текущем хф пофиксить и накотить? сделать промежуточный ХФ с исправлением кривой выплат

@21xhipster мб посмособствуешь? а тут кроме барыг скоро никого не будет. Да, они не исчезнут, но хотя бы авторы начнут себя более менее чувствовать

@litrbooh  @On1x  поднимите вопрос среди делегатов на эту тему плиз, согласны ли будут шустро накатить ХФ чисто с изменением выплат

**Comment by Nemo1369** [comment]
@t3ran13 Not a problem. Moreover, I have had these changes done (votable curve) in some private branch. The only trouble is the consensus.

**Comment by yudina-cat** [comment]
Я выгрузила данные по author rewards c 1 по 15 декабря, и по ним видно, что медианная выплата за месяц (с начала ноября) упала в 6 раз, до совсем печальных 0,21 GBG. @t3ran13 прав, нужно хотя бы с помощью сглаживания функции попробовать изменить ситуацию.

**Comment by litrbooh** [comment]
@Nemo1369 , проблема конценсуса только одна. Вы не приходите и не общаетесь с делегатами. Я уверен, что уже можно было бы прийти к консенсусу, если люди от ядра и КФ приходили на встречи делегатов и разговаривали не в стиле "мы решили, мы хотим". Все делегаты люди адекватные и всегда готовы искать компромисс, просто его надо искать.

Вопрос про изменение уже давно поднят, в целом консенсус о том, что её надо менять есть. Осталось понять на что. А это уже обсуждать надо, а никто из вас и КФ не хочет это обсуждать.

**Comment by litrbooh** [comment]
@Nemo1369 , как вариант - сделайте просто линейную функцию пока. А потом в следующем ХФ сделайте 2-3 на выбор, и пусть делегаты выбирают голосованием в сети уже.

**Comment by t3ran13** [comment]
@Nemo1369

я правильно понял, что у тебя все готово,  тебе просто нужно задать функцию?
а под консенсусом ты понимаешь единогласное мнение делегатов по функции??

**Comment by Nemo1369** [comment]
@t3ran13 That is what I’m talking about. But the requirements are to set up the consensus logarithmic functions set. (P.E. log10(x), log2(x), ln(x), xlog10(x), xlog2(x), xln(x)). Then I will be able to publish the branch.

С уважением,

Михаил Комаров
nemo1369@gmail.com

**Comment by gropox** [comment]
Я думаю большинство делегатов, за смену кривой. Так что думаю консенсус будет. 

Я за выпуск ХФ 0.16.5 в ближайшее время. Без майнеров (в пользу бэкап делегатов) и с линейной кривой. В 0.2.0, в или 0.2.1 если понадобится можно сменить кривую. Либо сделать параметрируемую делегатами имплементировав основные направления.

**Comment by t3ran13** [comment]
> В 0.2.0, в или 0.2.1 если понадобится можно сменить кривую. 
+1
начать с чего-то следует! а то смотреть как корабль тонет и никто ничего не спешит делать...

**Comment by litrbooh** [comment]
@Nemo1369 никто из делегатов не предлагал логарифм, никто не предагал аналитику ни одной из тех функций, которые вы тут назвали. Зато про линейную было много постов в том числе от нескольких ТОП делегатов.

**Comment by t3ran13** [comment]
@litrbooh  
как ты уже залолбал (тут должен быть мат) со своим миром грез связанным с линейностью.
вот скажи, @on1x топовый делегат? он предлагал какой вариант? прямо в этой же теме вот те ссылка
https://github.com/GolosChain/golos/issues/283#issuecomment-346309649

не говоря о том. что на самом голосе об этом варианте  говорили милион раз.
зачем сводить все либо только к линености либо только к квадрату? и там и там есть проблемы, чем тебя не устраивает средний вариант между ними?

таже @yudina-cat тоже делегат и сравнила на голосе логарифм и линейность

**Comment by litrbooh** [comment]
@t3ran13
Заколебали люди, которые 2 месяца назад кричали, что менять кривую не надо. А теперь им видите ли надо, но не линейную. Ну, не линеную будешь опять в бустерам отдавать предпочтение только в меньшей степени.

**Comment by litrbooh** [comment]
@Nemo1369 , не хотите включить этот ищью в 16.5?

**Comment by bitfag** [comment]
@Nemo1369 please avoid formatting unrelated code, move it to separate commits next time.

```diff
-                                   STEEMIT_PAYOUT_EXTENSION_COST_PER_DAY.amount.value /
-                                   (60 * 60 * 24), SBD_SYMBOL_NAME);
+                                   STEEMIT_PAYOUT_EXTENSION_COST_PER_DAY.amount.value / (60 * 60 * 24),
+                                   SBD_SYMBOL_NAME);

-                    ((input_cost.amount.value * 60 * 60 * 24) /
-                     STEEMIT_PAYOUT_EXTENSION_COST_PER_DAY.amount.value));
+                    ((input_cost.amount.value * 60 * 60 * 24) / STEEMIT_PAYOUT_EXTENSION_COST_PER_DAY.amount.value));

-                push_virtual_operation(fill_order_operation<0, 17, 0>(order.seller, order.order_id, pays, receives, issuer_fees));
+                push_virtual_operation(
+                        fill_order_operation<0, 17, 0>(order.seller, order.order_id, pays, receives, issuer_fees));
```

**Comment by AntonBliznets** [comment]
https://golos.io/ru--golos/@goloscore/novosti-golos-core-tekhnicheskoe-zadanie-na-modelirovanie-ekonomiki-golos
